### PR TITLE
add new static analysis for PHPMD tool

### PR DESCRIPTION
**Pull request by adirkuhn**
adding support for the text report for PHPMD

### CODE REVIEW

**Comment by DaveLiddament** [comment]
@adirkuhn Thanks for this PR. I shall have a look first thing tomorrow and aim to get it into a release by the end of the day. Thanks for the contribution.

**Comment by DaveLiddament** [comment]
Sorry for the delay in this. I was away all weekend. The code you've written is great. However there is a limitation with the PMD text output that means, in some contexts, the output from SARB will be misleading... 

Running PMD on SARB one of the errors I get is this:

```
/vagrant/tests/Unit/Plugins/GitDiffHistoryAnalyser/internal/OriginalLineNameCalculatorTest.php:84	Avoid using static access to class '\DaveLiddament\StaticAnalysisResultsBaseliner\Plugins\GitDiffHistoryAnalyser\internal\OriginalLineNumberCalculator' in method 'testLineNumber0'.
```

The code it refers to is the first line of this method...
```
    public function testLineNumber0(): void
    {
        $originalLineNumber = OriginalLineNumberCalculator::calculateOriginalLineNumber($this->fileMutation, 0);
        $this->assertSame(0, $originalLineNumber);
    }
```

SARB tracks the issue types to a line of code. [Full details](https://github.com/DaveLiddament/sarb/blob/master/docs/HowSarbWorks.md)
In this PR, the issue type is the error message. Given the text output from PMD, there is no alternative. 

Suppose I renamed the method from `testLineNumber0` to `testLine0`. 

The issue highlighted above hasn't changed. However the message for the issue (which is used as the issue type) will change (because the message contains the name of the method). SARB will consider this issue to be a new issue and incorrectly report it as a new issue, rather than an issue in the baseline. 


We've got 2 options...

1. Continue using the PMD text format. If we want to continue with this format then we should be aware there will be times when certain refactors might cause issues that should be in the baseline to be reported as new issues. 
2. Switch to a different PMD output format. I'd recommend JSON as it's easy to parse. I have time on Tuesday to create a PMD parser for that. 

The JSON format gives following information for each issue: 
```
        {
            "file": "\/vagrant\/src\/Plugins\/GitDiffHistoryAnalyser\/internal\/OriginalLineNumberCalculator.php",
            "violations": [
                {
                    "beginLine": 84,
                    "endLine": 84,
                    "package": null,
                    "function": null,
                    "class": null,
                    "method": null,
                    "description": "Avoid using static access to class \u0027\\DaveLiddament\\StaticAnalysisResultsBaseliner\\Plugins\\GitDiffHistoryAnalyser\\internal\\OriginalLineNumberCalculator\u0027 in method \u0027testLineNumber0\u0027.",
                    "rule": "StaticAccess",
                    "ruleSet": "Clean Code Rules",
                    "externalInfoUrl": "https:\/\/phpmd.org\/rules\/cleancode.html#staticaccess",
                    "priority": 1
                }
            ]
        },
```

With this information the issue type would be the `rule`.

File: src/Plugins/PhpMdTextResultsParser/PhpMdTextResultsParser.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    public function getIdentifier(): Identifier
+    {
+        return new PhpMdTextIdentifier();
+    }
+
+    public function showTypeGuessingWarning(): bool
+    {
+        return false;
```
Comment (DaveLiddament) [inline src/Plugins/PhpMdTextResultsParser/PhpMdTextResultsParser.php:34]: This should return true. As the type guessing is used. See: https://github.com/DaveLiddament/sarb/blob/master/docs/ViolationTypeClassificationGuessing.md

**Comment by adirkuhn** [comment]
hmm, I wasn't expecting this, thanks for the explanation. Anyway, I think we can move this PR to rejected and I can create a new PR with a JSON format. I think that will not possible for me this week, maybe the next one.

**Comment by DaveLiddament** [comment]
@adirkuhn I've created a PR #65 that can work with JSON output from PHP-MD. I'll release it shortly. 

Thank you so much for taking the time to make a PR for SARB. SARB is my first open source project and I can not describe how happy it makes me when people raise PRs on the project. Even though we've not been able to use this PR I am incredibly grateful for the interest and effort you've put in. 

If there are any other features that you think SARB would benefit from, or other tools you want support for then please let me know. Thanks very much.

**Comment by DaveLiddament** [comment]
@adirkuhn checkout release [0.8.0](https://github.com/DaveLiddament/sarb/releases/tag/0.8.0) for support with PHM - JSON format.

**Pull request closed without merging**